packages/flutter/lib/src/gestures/drag_boundary.dart

```diff
@@ -0,0 +1,130 @@
+  /// is within the boundary.
+  ///
+  /// Returns `true` if the location is within the boundary, and `false`
+  /// otherwise.
+  bool isWithinBoundary(Offset globalLocation);
+
+  /// Given a global location that is outside the boundary, finds and
+  /// returns the nearest position within the boundary.
+  ///
+  /// Returns an [Offset] representing the nearest position within the
+  /// boundary.
+  Offset getNearestPositionWithinBoundary(Offset globalLocation);
+}
+
+/// Defines a boundary for a drag point.
+///
+/// This class extends [DragBoundary] and is used to determine whether
+/// the currently dragged point is within the specified boundary.
+///
+/// The [isWithinBoundary] method checks if a given global location is
+/// within the boundary.
+///
+/// The [getNearestPositionWithinBoundary] method takes a global location
+/// that is outside the boundary and returns the nearest position within
+/// the boundary.
+class DragPointBoundary extends DragBoundary {
+  /// Creates a [DragPointBoundary] with the given [boundary].
+  DragPointBoundary(this.boundary);
+
+  /// The boundary of the drag gesture.
+  final Rect boundary;
+
+  @override
+  bool isWithinBoundary(Offset globalLocation) {
+    return boundary.contains(globalLocation);
+  }
+
+  @override
+  Offset getNearestPositionWithinBoundary(Offset globalLocation) {
+    final double dx = clampDouble(globalLocation.dx, boundary.left, boundary.right);
+    final double dy = clampDouble(globalLocation.dy, boundary.top, boundary.bottom);
+    return Offset(dx, dy);
+  }
+}
+
+/// Defines a boundary for a drag rectangle.
+///
+/// This class extends [DragBoundary] and is used to determine whether
+/// the rectangle being dragged is within the specified boundary.
+///
+/// The rectangle is defined by its top left corner (the drag point)
+/// and its size.
+///
+/// The [isWithinBoundary] method checks if the entire rectangle is
+/// within the boundary.
+///
+/// The [getNearestPositionWithinBoundary] method takes a global location
+/// that is outside the boundary and returns the nearest position within
+/// the boundary that the rectangle can be dragged to.
+class DragRectBoundary extends DragBoundary {
+  /// Creates a [DragRectBoundary] with the given [boundary],
+  /// [rectOffset], and [rectSize].
+  ///
+  /// [boundary] defines the boundary of the drag gesture. [rectOffset]
+  /// is the offset of the top left corner of the dragged rectangle
+  /// from the drag point. [rectSize] is the size of the dragged rectangle.
+  DragRectBoundary({
+    required this.boundary,
+    required this.rectOffset,
+    required this.rectSize,
+  });
+
+  /// Defines the boundary within which the drag gesture should
+  /// be contained.
+  final Rect boundary;
+
+  /// Represents the offset of the top left corner of the dragged rectangle
+  /// from the drag point. This is used to calculate the actual position
+  /// of the rectangle.
+  final Offset rectOffset;
```

Comment:
The documentation here was a little confusing on my first read since `DragRectBoundaryProvider` sets this value to the initial drag position. Maybe we can rename this field to `initialDragPosition` or explain more about the relation between the top-left and the initial drag position in an example. I think explaining what this value would be when the initial drag position is actually in the top left (`Offset.zero`) vs when it is inside the rectangle would be useful.